Public/Src/Engine/Dll/Distribution/RemoteWorker.cs

```diff
@@ -381,58 +381,112 @@ public override async void Start()
         /// <inheritdoc />
         public override async Task FinishAsync(string buildFailure)
         {
-            m_buildRequests.CompleteAdding();
-            if (m_sendThread.IsAlive)
+            while (true)
             {
-                m_sendThread.Join();
+                WorkerNodeStatus status = Status;
+                if (status == WorkerNodeStatus.Stopping || status == WorkerNodeStatus.Stopped)
+                {
+                    // We already initiated the stop for the worker.
+                    return;
+                }
+
+                if (ChangeStatus(status, WorkerNodeStatus.Stopping))
+                {
+                    break;
+                }
             }
 
-            bool initiatedStop = false;
+            await DisconnectAsync(buildFailure);
+        }
+
+        /// <inheritdoc />
+        public override async Task EarlyReleaseAsync()
+        {
+            // Unblock scheduler
+            await Task.Yield();
+
             while (true)
             {
                 WorkerNodeStatus status = Status;
                 if (status == WorkerNodeStatus.Stopping || status == WorkerNodeStatus.Stopped)
                 {
-                    break;
+                    // Already stopped, no need to continue.
+                    return;
                 }
 
                 if (ChangeStatus(status, WorkerNodeStatus.Stopping))
                 {
-                    initiatedStop = true;
                     break;
                 }
             }
 
-            if (initiatedStop)
-            {
-                CancellationTokenSource exitCancellation = new CancellationTokenSource();
+            var drainStopwatch = new StopwatchVar();
 
-                // Only wait a short amount of time for exit (15 seconds) if worker is not successfully attached.
-                if (m_attachCompletion.Task.Status != TaskStatus.RanToCompletion || !await m_attachCompletion.Task)
+            using (drainStopwatch.Start())
+            {
+                if (AcquiredSlots != 0)
                 {
-                    exitCancellation.CancelAfter(TimeSpan.FromSeconds(15));
+                    await DrainCompletion.Task;

```

Comment:
>await DrainCompletion.Task; [](start = 20, length = 27)

Why not always await this?

Reply:
ADded the reason as a comment

---
In reply to: [289427814](https://github.com/microsoft/BuildXL/pull/392#discussion_r289427814) [](ancestors = 289427814)